### ISSUE TITLE
[vioscsi] ConfigInfo->NumberOfPhysicalBreaks should be dependent on adaptExt->scsi_config.seg_max advertised by qemu

### ISSUE THREAD

**Issue by albertlav** [virtio-win/kvm-guest-drivers-windows#617]
As per Microsoft doc NumberOfPhysicalBreaks must be set according to HBA's scatter/gather capacity:

https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/srb/ns-srb-_port_configuration_information

> Specifies the maximum number of breaks between address ranges that a data buffer can have if the HBA supports scatter/gather. In other words, the number of scatter/gather lists minus one. By default, the value of this member is SP_UNINITIALIZED_VALUE, which indicates the HBA can support an unlimited number of physical discontiguities. If the port driver sets a value for this member, the miniport driver can adjust the value lower but no higher. If this member is SP_UNINITIALIZED_VALUE, the miniport driver must reset this member according to the HBA's scatter/gather capacity, with zero representing no scatter/gather support.
> 

If my understanding is correct HBA scatter/gather capacity advertised by QEMU in https://github.com/qemu/qemu/blob/master/hw/scsi/virtio-scsi.c#L741
as `scsiconf->seg_max` and this value visible to guest as `adaptExt->scsi_config.seg_max`  via `GetScsiConfig()`

However vioscsi currently set NumberOfPhysicalBreaks value based on registry or hardcoded value of 512

Probably NumberOfPhysicalBreaks  should depend on seg_max.

**Comment by tomty89** [comment]
In reality, from what I tested / experienced, NumberOfPhysicalBreaks cannot exceed `max_hw_sectors` of the backend drive. (The adjustable "soft limit" `max_sectors` did not seem to matter though.)

Say `/sys/block/$node/queue/max_hw_sectors*_kb*` is `512` (which is quite a common case). Then the `NumberOfPhysicalBreaks` (which is essentially a number of *pages*) cannot exceed `512 * 1024 = 524288` bytes, which it cannot be larger than `524288 / 4096 = 128` (in the typical 4KiB page size case). (Due to the "+1" requirement, whatever it really is, apparently the current hard-coded value `65` is maximum we can / would want to use for such drive. If for example / some reason the drive has `max_hw_sectors_kb` with a lower value like `256`, NOPB will need to be less than `64`, in which case the "optimal value" appears to be `33`.)

I've proposed to use the option `max_sectors` (of `-device virtio-scsi-pci`) in qemu as a *clamp* for NOPB. See https://github.com/virtio-win/kvm-guest-drivers-windows/pull/502/files for details, although it will never be merged as the dev(s) seems to be impossible to communicate with.
